TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Service/NodeTreeService.php

```diff
@@ -0,0 +1,52 @@
+<?php
+namespace TYPO3\TYPO3CR\Service;
+
+/*
+ * This file is part of the TYPO3.TYPO3CR package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use TYPO3\Flow\Annotations as Flow;
+use TYPO3\TYPO3CR\Domain\Factory\NodeFactory;
+use TYPO3\TYPO3CR\Domain\Model\NodeInterface;
+
+/**
+ * The node tree service
+ * @Flow\Scope("singleton")
+ */
+class NodeTreeService
+{
+
+    /**
+     * @Flow\Inject
+     * @var NodeFactory
+     */
+    protected $nodeFactory;
+
+
+    /**
+     * @param NodeInterface $node
+     * @param callable $action
+     * @return void
+     */
+    public function traverseTree(NodeInterface $node, callable $action)
+    {
+        $continue = $action($node);
+        if ($continue !== false && $node->hasChildNodes()) {
+            foreach ($node->getChildNodes() as $childNode) {
+                $this->traverseTree($childNode, $action);
+            }
+        }
+        $this->nodeFactory->removeNodeFromCache($node);
```

Comment:
> Would introducing something like disableCache() and enableCache() in NodeFactory and ContentContext be okay?
 
Don't do that, i prefer method that accept a callback ```\TYPO3\Flow\Security\Context::withoutAuthorizationChecks``` for a good example. Like this the API is clean, and the developper dont need to call the enable method after the processing and you can handle exception cleanly.

> a) With write/reset we then have n cache write operations, 0 cache hits and n reset operations for n nodes. Though everything happens in-memory and thus is fast, it's unnecessary none the less

I agree that if the tree traversal is done correctly the cache is not needed. But agree also with the "problem" with node repair, but I think it should be easy to do some bench mark. I prefer low memory usage and a bit longer CPU time